policybot/pkg/testresults/testresults.go

```diff
@@ -0,0 +1,337 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * Take in a pr number from path "istio-prow/pr-logs/pull/istio-istio" and examine the pr
+ * for all tests that are run and their results. The results are then written to Spanner.
+ */
+package testresults
+
+import (
+	"context"
+	"encoding/json"
+
+	"io/ioutil"
+	"strconv"
+	"strings"
+	"time"
+
+	"cloud.google.com/go/storage"
+	"google.golang.org/api/iterator"
+
+	store "istio.io/bots/policybot/pkg/storage"
+	"istio.io/pkg/log"
+)
+
+/*
+ * Pull struct for the structure under refs/pulls in clone-records.json
+ */
+type pull struct {
+	Number int
+	Author string
+	Sha    string
+}
+
+/*
+ * Cmd struct for Commands object under clone-records.json
+ */
+type cmnd struct {
+	Command string
+	Output  string
+}
+
+/*
+ * Finished struct to store values for all fields in Finished.json
+ */
+type finished struct {
+	Timestamp int64
+	Passed    bool
+	Result    string
+}
+
+/*
+ * Clone_Record struct to store values for all fields in clone-records.json
+ */
+type cloneRecord struct {
+	Refs struct {
+		Org       string
+		Repo      string
+		BaseRef   string `json:"base_ref"`
+		BaseSha   string `json:"base_sha"`
+		Pulls     []pull
+		PathAlias string
+	}
+	Commands []cmnd
+	Failed   bool
+}
+
+/*
+ * Started struct to store values from started.json
+ */
+type started struct {
+	Timestamp int64
+}
+
+type PrResultTester struct {
+	Client     *storage.Client
+	ctx        context.Context
+	bucketName string
+}
+
+var scope = log.RegisterScope("TestResult", "Check error while reading from google cloud storage", 0)
+
+/*
+ * Function to initionalize new PrFlakeTest object.
+ * Use `prFlakeyTest, err := NewPrFlakeTest()` and `testFlakes := prFlakeyTest.checkTestFlakesForPr(<pr_number>)`
+ * to get test flakes information for a given pr.
+ */
+func NewPrResultTester(ctx context.Context, bucketName string) (*PrResultTester, error) {
```

Comment:
Right now, this function is being called once per call to the filter. Would it make sense to create . this struct  just once when the bot starts up and simply reuse it for every call to the filter?